e2e/internal/keyexec/keyexec.go

```diff
@@ -0,0 +1,361 @@
+func PullDefaultPublicKey(t *testing.T) {
+	e2e.LoadEnv(t, &testenv)
+
+	argv := []string{"key", "pull", "F69C21F759C8EA06FD32CCF4536523CE1E109AF3"}
+
+	execKey := exec.Command(testenv.CmdPath, argv...)
+
+	out, err := execKey.CombinedOutput()
+	if err != nil {
+		t.Log(string(out))
+		t.Fatalf("Unable to pull key: %v", err)
+	}
+}
+
+// RemoveDefaultPublicKey will pull the public Sylabs Admin key
+func RemoveDefaultPublicKey(t *testing.T) {
+	e2e.LoadEnv(t, &testenv)
+
+	argv := []string{"key", "remove", "F69C21F759C8EA06FD32CCF4536523CE1E109AF3"}
+	execKey := exec.Command(testenv.CmdPath, argv...)
+
+	out, err := execKey.CombinedOutput()
+	if err != nil {
+		t.Log(string(out))
+		t.Fatalf("Unable to pull key: %v", err)
+	}
+}
+
+func RemovePublicKeyring(t *testing.T) {
+	err := os.Remove(sypgp.PublicPath())
+	if err != nil {
+		t.Fatalf("Unable to remove public keyring: %v", err)
+	}
+}
+
+// RemoveSecretKeyring will delete your secret keyring.
+func RemoveSecretKeyring(t *testing.T) {
+	err := os.Remove(sypgp.SecretPath())
+	if err != nil {
+		t.Fatalf("Unable to remove secret keyring: %v", err)
+	}
+}
+
+func RemoveKeyring(t *testing.T) {
+	err := os.RemoveAll(sypgp.DirPath())
+	if err != nil {
+		t.Fatalf("Unable to remove keyring directory: %v", err)
+	}
+}
+
+// ImportKey will import a key from kpath.
+func ImportKey(t *testing.T, kpath string) ([]byte, error) {
+	e2e.LoadEnv(t, &testenv)
+
+	argv := []string{"key", "import", kpath}
+	execKey := exec.Command(testenv.CmdPath, argv...)
+
+	return execKey.CombinedOutput()
+}
+
+// ImportPrivateKey will take a private key file (kpath) and import it.
+func ImportPrivateKey(t *testing.T, kpath string) (string, []byte, error) {
+	e2e.LoadEnv(t, &testenv)
+
+	c, err := expect.NewConsole(expect.WithStdin(os.Stdin))
+	if err != nil {
+		panic(err)
+	}
+	defer c.Close()
+
+	exportCmd := []string{"key", "import", kpath}
+
+	cmd := exec.Command(testenv.CmdPath, exportCmd...)
+	cmd.Stdin = c.Tty()
+
+	buf := bytes.NewBuffer(nil)
+	cmd.Stderr = buf
+	cmd.Stdout = buf
+
+	go func() {
+		c.ExpectEOF()
+	}()
+
+	err = cmd.Start()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Send the passcode to singularity
+	c.Send("e2etests\n")
+	c.Send("e2etests\n")
+	c.Send("e2etests\n")
+
+	err = cmd.Wait()
+	cm := fmt.Sprintf("%s %s", testenv.CmdPath, strings.Join(exportCmd, " "))
+
+	return cm, buf.Bytes(), err
+}
+
+// ExportPrivateKey will import a private key from kpath.
+func ExportPrivateKey(t *testing.T, kpath, num string, armor bool) (string, []byte, error) {
+	e2e.LoadEnv(t, &testenv)
+
+	c, err := expect.NewConsole(expect.WithStdin(os.Stdin))
+	if err != nil {
+		panic(err)
+	}
+	defer c.Close()
+
+	exportCmd := []string{"key", "export", "--secret"}
+
+	if armor {
+		exportCmd = append(exportCmd, "--armor")
+	}
+
+	exportCmd = append(exportCmd, kpath)
+
+	outErr := bytes.NewBuffer(nil)
+
+	cmd := exec.Command(testenv.CmdPath, exportCmd...)
+	cmd.Stdin = c.Tty()
+
+	cmd.Stderr = outErr
+	cmd.Stdout = outErr
```

Comment:
why are these 2 sent as error and output with the same return value?

Reply:
This is to capture the stdout, and stderr, and returns it. Useful for debugging or when a test fails.

Reply:
why not using the combined output instead?

Reply:
Because `cmd.CombinedOutput` will wait for the command to finish running, and we want to use `c.Send()` to send input for the command thats running, (we cant do that if the command is already done running). So here we use `cmd.Start()` to start the command, when send the input, and then wait for the command to finish running by `cmd.Wait()`. *hope that makes some sense...*